forwards_paper.tex

```diff
@@ -903,59 +903,80 @@ \subsubsection*{Estimates of run-time complexity}
+NumPy's array interface provides a great deal of flexibility and efficiency,
 and makes it straightforward to transfer data from sources
 such as HDF5 \citep{hdf5}or Dask~\citep{dask}.
-For small scale data and debugging purposes, a simple text format is also supported.
-
-% PLR: this was mentioned above, and belongs in the documentation, I think.
-% The API also provides functions to sort tables
-% to ensure that the records are in the form required to input
-% into an \msprime{} tree sequence object. (Simplification may also be required.)
-
-The \msprime{} Tables API is designed to be easily ``plugged'' into any simulation engine.
-One simply needs to fill the relevant columns,
-and to send them to \msprime{} at regular intervals for simplification.
-For example, in the benchmarking simulations above,
-we used the \fwdpp{} \cpp{} API \citep{fwdpp}
-to fill arrays of simple structures representing the node and edge data.
-Doing so requires no modification of the \fwdpp{} code base.
-Rather, we simply need to bookkeep parent/offspring labels,
-and perform simple processing of the recombination breakpoints from each mating event.
-Using \texttt{pybind11} (\url{https://github.com/pybind/pybind11/}),
-the node/edge arrays on the \cpp{} side are made visible, without copy, to Python as a multi-column numpy array,
-which is processed by the \msprime{} tables API.
-Our implementation also uses \texttt{pybind11}
-to create a Python package extending \texttt{fwdpy11}
-(a Python package based on \fwdpp{}; \url{https://github.com/molpopgen/fwdpy11}),
-% (A neat aside, that we may wish to skip here, is that it may be possible to use the tables
-% API from a command-line C/\cpp{} application via an embedded Python interpreter.)
+For small scale data and debugging purposes, a simple text based format is also supported.
+
+The \msprime\ Python Tables API provides a general purpose toolkit for importing
+and processing succinct tree sequences. Interoperation with Python simulators
+is then straightforward. For the proof-of-concept implementations discussed
+here, we took two different approaches to doing this. In the first method,
+based on \simupop{}, the simulation engine invokes callbacks at critical points
+of the simulation, and we infer nodes and edges from the information that is
+provided. Rows are appended to the tables one-by-one, and the tables are
+periodically sorted and simplified to control memory usage. The second method we developed uses
+\texttt{pybind11} (\url{https://github.com/pybind/pybind11/}) to inferface
+with the \fwdpp{} \cpp{} API \citep{fwdpp}. No modifications were
+required to the \fwdpp{} code base; rather, we simply need to bookkeep parent/offspring labels,
+and perform simple processing of the recombination breakpoints from each mating
+event to generate node and edge data. This information is then periodically copied
+to the \msprime\ Tables API, where it is sorted and simplified.
```

Comment:
Some of it isn't even copied, I believe, due to zero-copy access mentioned above. @molpopgen can confirm

Reply:
It's still copied into msprime's memory --- the zero-copy Kevin's referring to is the memory view of C++ structs available in numpy.

Reply:
OK gotcha. A python user can in principle use a view of msprime data without a copy; such user can do the same with code from a forward simulator that uses the NumPy API. But getting data from fwd simulator into msprime will always involve a copy.

Reply:
Yes, exactly. Ideally, we'd like to close this gap between the forward simulator and the tables API.